pkg/controller/master/node/controller.go

```diff
@@ -0,0 +1,123 @@
+package node
+
+import (
+	ctlappsv1 "github.com/rancher/wrangler-api/pkg/generated/controllers/apps/v1"
+	ctlbatchv1 "github.com/rancher/wrangler-api/pkg/generated/controllers/batch/v1"
+	ctlcorev1 "github.com/rancher/wrangler-api/pkg/generated/controllers/core/v1"
+	batchv1 "k8s.io/api/batch/v1"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/client-go/tools/record"
+)
+
+// Handler
+type Handler struct {
+	podCache         ctlcorev1.PodCache
+	nodes            ctlcorev1.NodeClient
+	nodeCache        ctlcorev1.NodeCache
+	jobs             ctlbatchv1.JobClient
+	jobCache         ctlbatchv1.JobCache
+	statefulSets     ctlappsv1.StatefulSetClient
+	statefulSetCache ctlappsv1.StatefulSetCache
+	recorder         record.EventRecorder
+}
+
+// OnNodeChanged automate the upgrade of node roles and balance of minio pods
+// If the number of masters in the cluster is less than spec number,
+// the harvester node with the smallest age is automatically promoted to be master.
+// If minio is deployed in distribute mode, make minio pods balanced.
+// Minio balancing only takes place when the node is ready and not in the upgrade state.
+func (h *Handler) OnNodeChanged(key string, node *corev1.Node) (*corev1.Node, error) {
+	if node == nil || node.DeletionTimestamp != nil {
+		return node, nil
+	}
+
+	nodeList, err := h.nodeCache.List(labels.Everything())
+	if err != nil {
+		return nil, err
+	}
+
+	promoteNodeList, err := h.selectPromoteNodeList(nodeList)
+	if err != nil {
+		return nil, err
+	}
+
+	switch len(promoteNodeList) {
+	case 0:
+		return h.makeMinioBalanced(key, node)
+	case 1:
+		return h.promote(promoteNodeList[0])
```

Comment:
Looks like this is handled in the default case so we probably don't need case 1.